### PR TITLE
Docker-compose implementation

### PR DESCRIPTION
**Pull request by joachimhgg**
**Description**

Move from `docker` to `docker-compose` in order to implement Theia IDE and Webots in separate containers. Each time a client is connecting, we get its content, add our own `docker-compose.yml`, create a `Dockerfile` for Webots if no Dockerfile is provided, build and launch the services provide in the yml via `docker-compose up`.

**Features**
- manage the opening or closing of multiples Webots instances with multiple services (test with 20 running dockers) whatever the status of the images (connecting, starting, Webots loaded, etc)
- clean unused and dangling images, volumes, networks, intermediate containers and containers older.
- keep the pulled images to not have to pull them again, example: 
image
- the simulation_server.py still work with `docker:false` config

**Documentation**

[Setup-a-Webots-Project-Repository](https://github.com/cyberbotics/webots-cloud/wiki/Setup-a-Webots-Project-Repository)
[Running-Webots-in-a-Docker-container](https://github.com/cyberbotics/webots-cloud/wiki/Running-Webots-in-a-Docker-container)

**Possible improvement**
- maximum of created networks: 30. It is corresponding to 30 docker-compose running at the same time, so it should not be a problem, just to keep in mind. Possible solution: [user-defined bridge networks](https://docs.docker.com/network/network-tutorial-standalone/#use-user-defined-bridge-networks)
- update to compose V2: https://docs.docker.com/compose/cli-command/. Keep the v1 at the moment as it seems more stable.
- user can add (through their Dockerfile) their own images and they won't be removed if not use. We could think about removing an image if not used in the last month.

**Security issues**
See #4273

**Issues**
 issues with ssh connection on the server. Not sure if it is due to this PR or to the server configuration.
`xauth: (argv):1:  unable to read any entries from file "(stdin)"`

**Usefull CLI to debug the server**
To inspect:

- `ps aux | grep simulation`
- `cd /tmp/webots/instances/`
- `docker ps`
- `docker network ls`
- `docker volume ls`
- `docker images`
- `docker exec -it 140067974464896_webots_1 /bin/bash`

To clean the docker server:
- `docker volume prune`
- `docker network prune`
- `docker system prune`

### CODE REVIEW

**Comment by joachimhgg** [comment]
> In https://github.com/cyberbotics/webots-cloud/wiki/Setup-a-Webots-Project-Repository:
> 
>     * The Dockerfile.default, docker-compose-default.yml and docker-compose-theia.yml links have no targets.
> 
>     * `The competitor controller are then started by Webots inside the Webots container of the competitor`: I am not sure to understand, is it the Webots container or the Docker container?
> 
> 
> Otherwise i seems good to me. I saw that you did a lot of modification on Docker but did you try the server without Docker to be sure that nothing was broken in the process?

* I will add the targets when the PR will be merged in `develop`.
* It is the Docker container of Webots, as a container is necessarily a Docker one.
* Yes the instances of Webots can still be created by changing the `docker` flag to `false` in the .json in the `config/simulation` folder.

The simulation server is now sending errors to the `server.js` which interrupt the starting of the simulation and display an appropriate alert depending on the error. The following cases are handle:
* if the world does not exist
* if the default dockerfile or docker-compose.yml are missing (it is a server issue)
* if Webots was unable to start in its container
* if the image is not available on cyberbotics's dockerHub 

In addition, if the user is using its own Dockerfile, it can visualize its building in the JavaScript console.

**Review by BenjaminDeleze — DISMISSED** [review]
Looks good to me, but as I am not familiar with Docker it is maybe good to have a second reviewer.

File: resources/web/server/simulation_server.py [hunk 1]
```diff
@@ -244,71 +244,131 @@ def runWebotsInThread(client):
...
+                    envVarDocker["DISPLAY"] = display
+                    envVarDocker["XAUTHORITY"] = xauth
+
+                config['dockerConfDir'] = config['webotsHome'] + '/resources/web/server/config/simulation/docker'
+                # create a Dockerfile if not provided in the project folder
+                defaultDockerfilePath = ''
+                if not os.path.isfile('Dockerfile'):
+                    defaultDockerfilePath = config['dockerConfDir'] + '/Dockerfile.default'
+                    if os.path.exists(defaultDockerfilePath):
+                        os.system(f'cp {defaultDockerfilePath} ./Dockerfile')
```
Comment (omichel) [inline resources/web/server/simulation_server.py:298]:
```suggestion
                        os.system(f'ln -s {defaultDockerfilePath} ./Dockerfile')
```

File: resources/web/server/simulation_server.py [hunk 2]
```diff
@@ -244,71 +244,131 @@ def runWebotsInThread(client):
...
+                            volume = line.split(':')[1]
+                            dockerComposePath = config['dockerConfDir'] + "/docker-compose-theia.yml"
+                            envVarDocker["THEIA_V"] = volume
+                            envVarDocker["THEIA_PORT"] = port + 500
+
+                if not os.path.exists(dockerComposePath):
+                    dockerComposePath = config['dockerConfDir'] + "/docker-compose-default.yml"
+
+                if os.path.exists(dockerComposePath):
+                    os.system(f'cp {dockerComposePath} ./docker-compose.yml')
```
Comment (omichel) [inline resources/web/server/simulation_server.py:321]:
```suggestion
                    os.system(f'ln -s {dockerComposePath} ./docker-compose.yml')
```

File: resources/web/server/simulation_server.py [hunk 3]
```diff
@@ -244,71 +244,131 @@ def runWebotsInThread(client):
...
                 else:
-                    command += ' --gpus=all -e DISPLAY'
+                    error = f"error: Miss docker-compose-default.yml in {config['dockerConfDir']}"
```
Comment (omichel) [inline resources/web/server/simulation_server.py:323]:
```suggestion
                    error = f"Error: Missing docker-compose-default.yml in {config['dockerConfDir']}"
```

File: resources/web/server/simulation_server.py [hunk 4]
```diff
@@ -337,19 +398,32 @@ def on_exit(self):
...
+                logging.warning(f'[{id(self)}] Webots [{self.webots_process.pid}] was killed')
+                if sys.platform == 'darwin':
                     self.webots_process.kill()
-            self.webots_process = None
+                else:
+                    self.webots_process.terminate()
+                    try:
+                        self.webots_process.wait(5)  # set a timeout (seconds) to avoid blocking the whole script
+                    except subprocess.TimeoutExpired:
+                        logging.warning(f'[{id(self)}] ERROR killing Webots [{self.webots_process.pid}]')
```
Comment (omichel) [inline resources/web/server/simulation_server.py:424]:
```suggestion
                        logging.warning(f'[{id(self)}] Error killing Webots [{self.webots_process.pid}]')
```

File: resources/web/server/config/simulation/docker/docker-compose-theia.yml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      - NVIDIA_DRIVER_CAPABILITIES=all
+    ports:
+      - '$PORT:$PORT'
+    privileged: true
+    stdin_open: true # docker run -i
+    tty: true # docker run -t
+    volumes:
+      - ${XAUTHORITY:-/dev/null}:${XAUTHORITY:-/dev/null}
+      - /tmp/.X11-unix:/tmp/.X11-unix:rw
+      - controller_data:/usr/local/$THEIA_V:rw
```
Comment (omichel) [inline resources/web/server/config/simulation/docker/docker-compose-theia.yml:36]:
```suggestion
      - controller_data:/usr/local/$THEIA_VOLUME:rw
```

File: resources/web/server/simulation_server.py [hunk 6]
```diff
@@ -244,71 +244,131 @@ def runWebotsInThread(client):
...
+                # create a docker-compose.yml
+                dockerComposePath = ''
+                if os.path.exists('webots.yml'):
+                    with open('webots.yml', 'r') as webotsYml_file:
+                        data = webotsYml_file.read().splitlines(True)
+                    for line in data:
+                        if line.startswith("theia:"):
+                            volume = line.split(':')[1]
+                            dockerComposePath = config['dockerConfDir'] + "/docker-compose-theia.yml"
+                            envVarDocker["THEIA_V"] = volume
```
Comment (omichel) [inline resources/web/server/simulation_server.py:314]:
```suggestion
                            envVarDocker["THEIA_VOLUME"] = volume
```

**Review by omichel — CHANGES_REQUESTED** [review]
It looks pretty good to me (although I didn't tested).

File: resources/web/server/config/simulation/docker/docker-compose-theia.yml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+  theia:
+    image: joachimhgg/theiatest:test1
```
[the remark this answers is not in the document]
Comment (joachimhgg) [reply resources/web/server/config/simulation/docker/docker-compose-theia.yml:39]: Theia does not provide official images anymore, so we will have the need to build it, depending of our needs. For the moment, I have just built this one for test.

File: resources/web/server/simulation_server.py [hunk 8]
```diff
@@ -244,71 +244,131 @@ def runWebotsInThread(client):
             global config
             world = f'{self.project_instance_path}/worlds/{self.world}'
             port = client.streaming_server_port
+            asyncio.set_event_loop(asyncio.new_event_loop())
+            if not os.path.exists(world):
+                error = f"error: {self.world} does not exist."
```
[the remark this answers is not in the document]
Comment (joachimhgg) [reply resources/web/server/simulation_server.py:249]: This error are sent to the `server.js`, hence I kept the same syntax that we can find on the function `onMessage(event)`

**Review by omichel — DISMISSED** [review]
Thank you.

**Pull request merged**